keps/sig-network/2595-expanded-dns-config/README.md

```diff
@@ -0,0 +1,283 @@
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Allow kubernetes to have expanded DNS(Domain Name System) configuration.
+
+## Motivation
+
+Kubernetes today limits DNS configuration according to [the obsolete
+criteria](https://access.redhat.com/solutions/58028). As recent DNS resolvers
+allow an arbitrary number of search paths, a new feature gate
+`ExpandedDNSConfig` will be introduced. With this feature, kubernetes allows
+more DNS search paths and longer list of DNS search paths to keep up with recent
+DNS resolvers.
+
+Confirmed that expanded DNS configuration is supported by
+- `glibc 2.17-323`
+- `glibc 2.28`
+- `musl libc 1.22`
+- `pure Go 1.10 resolver`
+- `pure Go 1.16 resolver`
+
+### Goals
+
+- Make `kube-apiserver` allow expanded DNS configuration when validating Pod's
+or PodTemplate's `DNSConfig`
+- Make `kubelet` allow expanded DNS configuration when validating `resolvConf`
+- Make `kubelet` allow expanded DNS configuration when validating actual DNS
+resolver configuration composed of `cluster domain suffixes`(e.g.
+    default.svc.cluster.local, svc.cluster.local, cluster.local), kubelet's
+`resolvConf` and Pod's `DNSConfig`
+
+### Non-Goals
+
+- Remove limitation on DNS search paths completely
+- Let cluster administrators limit the number of search paths or the length of
+DNS search path list to an arbitrary number
+
+## Proposal
+
+- Expand `MaxDNSSearchPaths` to 32
+- Expand `MaxDNSSearchListChars` to 2048
+
+### User Stories (Optional)
+
+#### Story 1
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+This enhancement relaxes the validation of `Pod` and `PodTemplate`. Once the
+feature is activated, it must be carefully disabled. Otherwise, the objects left
+over from the previous version which have the expanded DNS configuration could
+be problematic.
+
+### Risks and Mitigations
+
+There may be some environments(DNS resolver or others) that break without
+current limitations. At this point, it is fair to open a bug, so they can fix it.
+
+## Design Details
+
+- Declare and define `MaxDNSSearchPathsExpanded` to `32` and
+`MaxDNSSearchListCharsExpanded` to `2048`
+- Add the feature gate `ExpandedDNSConfig` (see [Feature Enablement and
+Rollback](#feature-enablement-and-rollback))
+- If the feature gate `ExpandedDNSConfig` is enabled, replace
+`MaxDNSSearchPaths` with `MaxDNSSearchPathsExpanded` and replace
+`MaxDNSSearchListChars` with `MaxDNSSearchListCharsExpanded` to allow expanded
+DNS configuration
+
+### Test Plan
+
+- Add unit tests of validating expanded DNS config
+
+### Graduation Criteria
+
+#### Alpha -> Beta Graduation
+
+- Address feedback from alpha
+- Sufficient testing
+
+#### Beta -> GA Graduation
+
+- Address feedback from beta
+- Sufficient number of users using the feature
+- Close on any remaining open issues & bugs
+
+### Upgrade / Downgrade Strategy
+
+N/A
+
+### Version Skew Strategy
+
+In clusters with a replicated control plane, all kube-apiservers should enable
```

Comment:
This is more about kubelets.  It is valid to have kubelets be a (slightly) lower version than apiserver.  Looking at kubelet code I can see this will throw warnings (pkg/kubelet/network/dns/dns.go `Configurer.CheckLimitsForResolvConf()`) but that does not hard-fail.  Still worth noting here.

In the same file, `Configurer.formDNSSearchFitsLimits()`, it *does* truncate.  That is what is important to note here (and to test-confirm) -- back-rev kubelet should be fine, it will just truncate the results.

Reply:
Updated.

thanks @thockin